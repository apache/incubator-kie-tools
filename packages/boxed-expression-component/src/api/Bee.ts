/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { v4 as uuid } from "uuid";

export {};

/**
 * Generates a UUID with a format similar to _6EFDBCB4-F4AF-4E9A-9A66-2A9F24185674
 */
export const generateUuid = () => {
  return `_${uuid()}`.toLocaleUpperCase();
};
