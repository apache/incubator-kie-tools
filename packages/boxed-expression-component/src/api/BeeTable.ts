/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import * as React from "react";
import * as ReactTable from "react-table";
import { ResizerStopBehavior } from "../resizing/ResizingWidthsContext";
import { BeeTableCellUpdate, BeeTableColumnUpdate } from "../table/BeeTable/BeeTableHeader";
import { BeeTableSelection } from "../selection/BeeTableSelectionContext";
import { FeelVariables } from "@kie-tools/dmn-feel-antlr4-parser";

export interface BeeTableCellProps<R extends object> {
  data: readonly R[];
  rowIndex: number;
  columnIndex: number;
  columnId: string;
}

export interface BeeTableProps<R extends object> {
  /** Table identifier, useful for nested structures */
  tableId?: string;
  /** Optional children element to be appended below the table content */
  additionalRow?: React.ReactElement[];
  /** Optional label, that may depend on column, to be used for the popover that appears when clicking on column header */
  editColumnLabel?: string | { [groupType: string]: string };
  /** Option to enable or disable header edits */
  isEditableHeader?: boolean;
  /** Top-left cell custom content */
  controllerCell?: string | JSX.Element;
  //
  rowWrapper?: React.FunctionComponent<React.PropsWithChildren<{ row: R; rowIndex: number }>>;
  /** For each column there is a default component to be used to render the related cell */
  cellComponentByColumnAccessor?: { [columnId: string]: React.FunctionComponent<BeeTableCellProps<R>> };
  /** Table's columns */
  columns: ReactTable.Column<R>[];
  /** Table's cells */
  rows: R[];
  /** Function to be executed when cells are modified */
  onCellUpdates?: (cellUpdates: BeeTableCellUpdate<R>[]) => void;
  /** Function to be executed when columns are modified */
  onColumnUpdates?: (columnUpdates: BeeTableColumnUpdate<R>[]) => void;
  /** Custom configuration for the table context menu */
  operationConfig?: BeeTableOperationConfig;
  /** Collection of allowed operations that are allowed for current table 'conditions' */
  allowedOperations: (conditions: BeeTableContextMenuAllowedOperationsConditions) => BeeTableOperation[];
  /** The way in which the header will be rendered */
  headerVisibility?: BeeTableHeaderVisibility;
  /** Number of levels in the header, 0-based */
  headerLevelCountForAppendingRowIndexColumn?: number;
  /** True, for skipping the creation in the DOM of the last defined header group */
  skipLastHeaderGroup?: boolean;
  /** Custom function for getting row key prop, and avoid using the row index */
  getRowKey?: (row: ReactTable.Row<R>) => string;
  /** Custom function for getting column key prop, and avoid using the column index */
  getColumnKey?: (column: ReactTable.ColumnInstance<R>) => string;
  /** Disable/Enable cell edits. Enabled by default */
  isReadOnly?: boolean;
  /** Enable keyboard navigation */
  enableKeyboardNavigation?: boolean;
  /** */
  onRowAdded?: (args: { beforeIndex: number }) => void;
  onRowDuplicated?: (args: { rowIndex: number }) => void;
  onRowReset?: (args: { rowIndex: number }) => void;
  onRowDeleted?: (args: { rowIndex: number }) => void;
  onColumnAdded?: (args: { beforeIndex: number; groupType: string | undefined }) => void;
  onColumnDeleted?: (args: { columnIndex: number; groupType: string | undefined }) => void;
  shouldRenderRowIndexColumn: boolean;
  shouldShowRowsInlineControls: boolean;
  shouldShowColumnsInlineControls: boolean;
  resizerStopBehavior: ResizerStopBehavior;
  lastColumnMinWidth?: number;
  /** The variables for the current expression context */
  variables?: FeelVariables;
}

/** Possible status for the visibility of the Table's Header */
export enum BeeTableHeaderVisibility {
  AllLevels,
  LastLevel,
  SecondToLastLevel,
  None,
}

/** Table allowed operations */
export enum BeeTableOperation {
  ColumnInsertLeft,
  ColumnInsertRight,
  ColumnInsertN,
  ColumnDelete,
  RowInsertAbove,
  RowInsertBelow,
  RowInsertN,
  RowDelete,
  RowReset,
  RowDuplicate,
  SelectionCopy,
  SelectionCut,
  SelectionPaste,
  SelectionReset,
}

export interface BeeTableOperationGroup {
  /** Name of the group (localized) */
  group: string;
  /** Collection of operations belonging to this group */
  items: {
    /** Name of the operation (localized) */
    name: string;
    /** Type of the operation */
    type: BeeTableOperation;
  }[];
}

export type BeeTableOperationConfig =
  | BeeTableOperationGroup[]
  | { [columnGroupType: string]: BeeTableOperationGroup[] };

export type BeeTableContextMenuAllowedOperationsConditions = {
  selection: BeeTableSelection;
  column: ReactTable.ColumnInstance<any> | undefined;
  columns: ReactTable.ColumnInstance<any>[] | undefined;
};
