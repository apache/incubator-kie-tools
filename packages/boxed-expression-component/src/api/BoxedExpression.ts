/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  DMN15__tConditional,
  DMN15__tContext,
  DMN15__tDecisionTable,
  DMN15__tFilter,
  DMN15__tFor,
  DMN15__tFunctionDefinition,
  DMN15__tInformationItem,
  DMN15__tInvocation,
  DMN15__tList,
  DMN15__tLiteralExpression,
  DMN15__tQuantified,
  DMN15__tRelation,
} from "@kie-tools/dmn-marshaller/dist/schemas/dmn-1_5/ts-gen/types";

export type BoxedLiteral = { __$$element: "literalExpression" } & DMN15__tLiteralExpression;
export type BoxedRelation = { __$$element: "relation" } & DMN15__tRelation;
export type BoxedContext = { __$$element: "context" } & DMN15__tContext;
export type BoxedDecisionTable = { __$$element: "decisionTable" } & DMN15__tDecisionTable;
export type BoxedList = { __$$element: "list" } & DMN15__tList;
export type BoxedInvocation = { __$$element: "invocation" } & DMN15__tInvocation;
export type BoxedFunction = { __$$element: "functionDefinition" } & DMN15__tFunctionDefinition;
export type BoxedFor = { __$$element: "for" } & DMN15__tFor;
export type BoxedEvery = { __$$element: "every" } & DMN15__tQuantified;
export type BoxedSome = { __$$element: "some" } & DMN15__tQuantified;
export type BoxedConditional = { __$$element: "conditional" } & DMN15__tConditional;
export type BoxedFilter = { __$$element: "filter" } & DMN15__tFilter;

export enum BoxedFunctionKind {
  Feel = "FEEL",
  Java = "Java",
  Pmml = "PMML",
}

export type BoxedIterator = BoxedFor | BoxedEvery | BoxedSome;

export type BoxedExpression =
  | BoxedLiteral
  | BoxedRelation
  | BoxedContext
  | BoxedDecisionTable
  | BoxedList
  | BoxedInvocation
  | BoxedFunction
  | BoxedFor
  | BoxedEvery
  | BoxedSome
  | BoxedConditional
  | BoxedFilter;

export interface PmmlDocument {
  document: string;
  modelsFromDocument?: {
    model: string;
    parametersFromModel?: DMN15__tInformationItem[];
  }[];
}

// Please find a better place for this function as part of https://github.com/kiegroup/kie-issues/issues/34
export const getNextAvailablePrefixedName = (
  names: string[],
  namePrefix: string,
  lastIndex: number = names.length
): string => {
  const candidate = `${namePrefix}-${lastIndex + 1}`;
  const elemWithCandidateName = names.indexOf(candidate);
  return elemWithCandidateName >= 0 ? getNextAvailablePrefixedName(names, namePrefix, lastIndex + 1) : candidate;
};
