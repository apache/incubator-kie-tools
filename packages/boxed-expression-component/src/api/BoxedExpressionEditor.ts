/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import _ from "lodash";
import {
  ContextProps,
  DecisionTableProps,
  ExpressionProps,
  FunctionProps,
  InvocationProps,
  ListProps,
  LiteralExpressionProps,
  RelationProps,
} from "./ExpressionProps";
import { v4 as uuid } from "uuid";

export {};

declare global {
  // Set of Functions used by the BoxedExpressionEditor and the BoxedExpressionWrapper
  interface BeeApi {
    resetExpressionDefinition: (definition: ExpressionProps) => void;
    broadcastLiteralExpressionDefinition: (definition: LiteralExpressionProps) => void;
    broadcastRelationExpressionDefinition: (definition: RelationProps) => void;
    broadcastContextExpressionDefinition: (definition: ContextProps) => void;
    broadcastListExpressionDefinition: (definition: ListProps) => void;
    broadcastInvocationExpressionDefinition: (definition: InvocationProps) => void;
    broadcastFunctionExpressionDefinition: (definition: FunctionProps) => void;
    broadcastDecisionTableExpressionDefinition: (definition: DecisionTableProps) => void;
    notifyUserAction: () => void;
    openManageDataType: () => void;
    onLogicTypeSelect: (selectedLogicType: string) => void;
    selectObject: (uuid?: string) => void;
  }

  //API that BoxedExpressionEditor (bee) and its wrapper are expecting to be defined in the Window namespace
  interface Window {
    beeApi: BeeApi;
    beeApiWrapper: BeeApi;
  }
}

/**
 * First, checks if each field, specified in propertiesToCheck, is equal for prevDef and updatedDef
 * Second, if the condition specified above is true, executes functionToExecute
 * @param prevDef previous definition
 * @param updatedDef updated definition
 * @param functionToExecute function to be executed if criteria is met
 * @param propertiesToCheck properties to be checked in the input objects
 */
export const executeIfExpressionDefinitionChanged = (
  prevDef: ExpressionProps,
  updatedDef: ExpressionProps,
  functionToExecute: () => void,
  propertiesToCheck?: string[]
) => {
  const customizer = propertiesToCheck
    ? (prev: Record<string, any>, next: Record<string, any>) =>
        _.every(propertiesToCheck, (prop) => _.isEqual(prev[prop], next[prop]))
    : undefined;

  if (!_.isEqualWith(prevDef, updatedDef, customizer)) {
    functionToExecute();
  }
};

/**
 * Generates an UUID with a format similar to _6EFDBCB4-F4AF-4E9A-9A66-2A9F24185674
 */
export const generateUuid = () => {
  return `_${uuid()}`.toLocaleUpperCase();
};
