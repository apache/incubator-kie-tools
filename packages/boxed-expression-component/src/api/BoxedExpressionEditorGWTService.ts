/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ContextProps,
  DecisionTableProps,
  ExpressionProps,
  FunctionProps,
  InvocationProps,
  ListProps,
  LiteralExpressionProps,
  RelationProps,
} from "./ExpressionProps";

/**
 * This interface defines all the API methods which BoxedExpressionEditor component can use to dialog with GWT Layer
 */
export interface BoxedExpressionEditorGWTService {
  resetExpressionDefinition: (definition: ExpressionProps) => void;
  broadcastLiteralExpressionDefinition: (definition: LiteralExpressionProps) => void;
  broadcastRelationExpressionDefinition: (definition: RelationProps) => void;
  broadcastContextExpressionDefinition: (definition: ContextProps) => void;
  broadcastListExpressionDefinition: (definition: ListProps) => void;
  broadcastInvocationExpressionDefinition: (definition: InvocationProps) => void;
  broadcastFunctionExpressionDefinition: (definition: FunctionProps) => void;
  broadcastDecisionTableExpressionDefinition: (definition: DecisionTableProps) => void;
  notifyUserAction: () => void;
  openManageDataType: () => void;
  onLogicTypeSelect: (selectedLogicType: string) => void;
  selectObject: (uuid?: string) => void;
}
