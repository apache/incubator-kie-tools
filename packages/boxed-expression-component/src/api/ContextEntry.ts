/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DataType } from "./DataType";
import { ExpressionProps } from "./ExpressionProps";
import * as _ from "lodash";
import { DataRecord, Row } from "react-table";
import { TableHandlerConfiguration, TableOperation } from "./Table";
import { BoxedExpressionEditorI18n } from "../i18n";

export interface EntryInfo {
  /** Entry id */
  id: string;
  /** Entry name */
  name: string;
  /** Entry data type */
  dataType: DataType;
}

export interface ContextEntryRecord<T = ExpressionProps> extends DataRecord {
  entryInfo: EntryInfo;
  /** Entry expression */
  entryExpression: T;
  /** True, for synchronizing name and dataType parameters, between entryInfo and entryExpression */
  nameAndDataTypeSynchronized?: boolean;
  /** Callback to be invoked on expression resetting */
  onExpressionResetting?: () => void;
}

export type ContextEntries<T = ExpressionProps> = ContextEntryRecord<T>[];

export const DEFAULT_ENTRY_INFO_MIN_WIDTH = 150;
export const DEFAULT_ENTRY_EXPRESSION_MIN_WIDTH = 370;

export const getHandlerConfiguration = (
  i18n: BoxedExpressionEditorI18n,
  groupName: string
): TableHandlerConfiguration => [
  {
    group: groupName,
    items: [
      { name: i18n.rowOperations.insertAbove, type: TableOperation.RowInsertAbove },
      { name: i18n.rowOperations.insertBelow, type: TableOperation.RowInsertBelow },
      { name: i18n.rowOperations.delete, type: TableOperation.RowDelete },
      { name: i18n.rowOperations.clear, type: TableOperation.RowClear },
    ],
  },
];

export const generateNextAvailableEntryName = (
  entryInfos: EntryInfo[],
  namePrefix: string,
  lastIndex: number = entryInfos.length
): string => {
  const candidateName = `${namePrefix}-${lastIndex === 0 ? 1 : lastIndex}`;
  const entryWithCandidateName = _.find(entryInfos, { name: candidateName });
  return entryWithCandidateName ? generateNextAvailableEntryName(entryInfos, namePrefix, lastIndex + 1) : candidateName;
};

export const getEntryKey = (row: Row): string => {
  const entryRecord = row.original as ContextEntryRecord;
  return entryRecord.entryInfo.id ?? (row.original as DataRecord).id;
};

export function resetEntry(row: ContextEntryRecord): ContextEntryRecord {
  return {
    ...row,
    entryExpression: { id: (row.entryExpression as ExpressionProps).id },
  };
}
