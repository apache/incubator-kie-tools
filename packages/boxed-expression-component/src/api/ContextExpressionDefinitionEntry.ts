/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DmnBuiltInDataType } from "./DmnBuiltInDataType";
import { ExpressionDefinition } from "./ExpressionDefinition";
import * as _ from "lodash";
import { BeeTableOperationHandlerConfig, BeeTableOperation } from "./BeeTable";
import { BoxedExpressionEditorI18n } from "../i18n";
import { ExpressionDefinitionLogicType } from "./ExpressionDefinitionLogicType";

export interface ContextExpressionDefinitionEntryInfo {
  /** Entry id */
  id: string;
  /** Entry name */
  name: string;
  /** Entry data type */
  dataType: DmnBuiltInDataType;
}

export interface ContextExpressionDefinitionEntry<T extends ExpressionDefinition = ExpressionDefinition> {
  entryInfo: ContextExpressionDefinitionEntryInfo;
  /** Entry expression */
  entryExpression: T;
  /** True, for synchronizing name and dataType parameters, between entryInfo and entryExpression */
  nameAndDataTypeSynchronized?: boolean;
}

export const CONTEXT_ENTRY_INFO_MIN_WIDTH = 150;
export const CONTEXT_ENTRY_EXPRESSION_MIN_WIDTH = 370;

export const getOperationHandlerConfig = (
  i18n: BoxedExpressionEditorI18n,
  groupName: string
): BeeTableOperationHandlerConfig => [
  {
    group: groupName,
    items: [
      { name: i18n.rowOperations.insertAbove, type: BeeTableOperation.RowInsertAbove },
      { name: i18n.rowOperations.insertBelow, type: BeeTableOperation.RowInsertBelow },
      { name: i18n.rowOperations.delete, type: BeeTableOperation.RowDelete },
      { name: i18n.rowOperations.clear, type: BeeTableOperation.RowClear },
    ],
  },
];

export const getNextAvailableContextExpressionEntryName = (
  entryInfos: ContextExpressionDefinitionEntryInfo[],
  namePrefix: string,
  lastIndex: number = entryInfos.length
): string => {
  const candidateName = `${namePrefix}-${lastIndex === 0 ? 1 : lastIndex}`;
  const entryWithCandidateName = _.find(entryInfos, { name: candidateName });
  return entryWithCandidateName
    ? getNextAvailableContextExpressionEntryName(entryInfos, namePrefix, lastIndex + 1)
    : candidateName;
};

export function resetContextExpressionEntry(entry: ContextExpressionDefinitionEntry): ContextExpressionDefinitionEntry {
  return {
    ...entry,
    entryExpression: {
      id: entry.entryExpression.id,
      logicType: ExpressionDefinitionLogicType.Undefined,
    },
  };
}
