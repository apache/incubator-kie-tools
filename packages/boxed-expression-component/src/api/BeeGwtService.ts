/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { BoxedExpression } from "./BoxedExpression";

/**
 * This interface defines all the API methods which BoxedExpressionEditor component can use to dialog with GWT Layer
 */
export interface BeeGwtService {
  getDefaultExpressionDefinition(
    logicType: BoxedExpression["__$$element"] | undefined,
    typeRef: string,
    isRoot?: boolean
  ): { expression: BoxedExpression; widthsById: Map<string, number[]> };

  openDataTypePage: () => void; // Just open the data types tab

  selectObject: (uuid?: string) => void; // Changes the Properties panel
}
