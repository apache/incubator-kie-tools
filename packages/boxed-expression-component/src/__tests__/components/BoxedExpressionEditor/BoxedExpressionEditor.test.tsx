/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { fireEvent, render } from "@testing-library/react";
import { BoxedExpressionEditor } from "../../../components";
import { DataType, ExpressionProps, LogicType } from "../../../api";
import { act } from "react-dom/test-utils";

describe("BoxedExpressionEditor tests", () => {
  test("should render BoxedExpressionEditor component", () => {
    const selectedExpression: ExpressionProps = { name: "Expression Name", dataType: DataType.Undefined };

    const { container } = render(<BoxedExpressionEditor expressionDefinition={selectedExpression} />);

    expect(container).toMatchSnapshot();
  });

  test("should reset the selection, when logic type is selected and clear button gets clicked", async () => {
    const expression = {
      name: "Test",
      logicType: LogicType.LiteralExpression,
      dataType: DataType.Undefined,
    };

    const { baseElement } = render(<BoxedExpressionEditor expressionDefinition={expression} />);

    fireEvent.contextMenu(baseElement.querySelector(".logic-type-selector") as HTMLElement);

    act(() => {
      const clearButtonElement = baseElement.querySelector(".context-menu-container button")!;
      const clearButton = clearButtonElement as HTMLButtonElement;
      clearButton.click();
    });

    expect(baseElement.querySelector(".logic-type-selector")).toBeTruthy();
    expect(baseElement.querySelector(".logic-type-selector")).toHaveClass("logic-type-not-present");
  });
});
