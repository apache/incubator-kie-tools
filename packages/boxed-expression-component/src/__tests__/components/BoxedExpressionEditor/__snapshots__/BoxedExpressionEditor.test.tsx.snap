// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxedExpressionEditor tests should render BoxedExpressionEditor component 1`] = `
<div>
  <div
    class="boxed-expression-editor"
  >
    <div
      class="expression-container"
    >
      <div
        class="expression-name-and-logic-type"
      >
        <span
          class="expression-title"
        >
          Expression Name
        </span>
        <span
          class="expression-type"
        >
          (
          &lt;Undefined&gt;
          )
        </span>
      </div>
      <div
        class="expression-container-box"
        data-ouia-component-id="expression-container"
      >
        <div
          class="logic-type-selector logic-type-not-present"
        >
          Select expression
        </div>
      </div>
    </div>
  </div>
  <div
    class="kie-cell-selection-box"
  >
    <div
      class="kie-selection-box"
    />
    <textarea />
  </div>
</div>
`;
